runtime/runtime/src/balance_checker.rs

```diff
@@ -476,4 +480,62 @@ mod tests {
             Err(RuntimeError::UnexpectedIntegerOverflow)
         );
     }
+
+    /// This tests shows what would happen if the total balance becomes u128::MAX
+    /// which is also the sentinel value use to distinguish between accounts version 1 and 2 or higher
+    /// See NEP-491 for more details: https://github.com/near/NEPs/pull/491.
+    #[test]
+    fn test_total_balance_u128_max() {
+        let tries = TestTriesBuilder::new().build();
+        let root = MerkleHash::default();
+        let alice_id = alice_account();
+        let bob_id = bob_account();
+        let gas_price = 100;
+        let deposit = 1000;
+
+        let mut initial_state = tries.new_trie_update(ShardUId::single_shard(), root);
+        let alice = account_new(u128::MAX - 1, hash(&[]));
+        let bob = account_new(1u128, hash(&[]));
+
+        set_account(&mut initial_state, alice_id.clone(), &alice);
+        set_account(&mut initial_state, bob_id.clone(), &bob);
+        initial_state.commit(StateChangeCause::NotWritableToDisk);
+
+        let signer =
+            InMemorySigner::from_seed(alice_id.clone(), KeyType::ED25519, alice_id.as_ref());
+
+        let tx =
+            SignedTransaction::send_money(0, alice_id, bob_id, &signer, 1, CryptoHash::default());
+
+        let receipt = Receipt {
+            predecessor_id: tx.transaction.signer_id.clone(),
+            receiver_id: tx.transaction.receiver_id.clone(),
+            receipt_id: Default::default(),
+            receipt: ReceiptEnum::Action(ActionReceipt {
+                signer_id: tx.transaction.signer_id.clone(),
+                signer_public_key: tx.transaction.public_key.clone(),
+                gas_price,
+                output_data_receivers: vec![],
+                input_data_ids: vec![],
+                actions: vec![Action::Transfer(TransferAction { deposit })],
+            }),
+        };
+
+        // Alice's balance becomes u128::MAX, which causes it is interpreted as
+        // the Alice's account version to be 2 or higher, instead of being interpreted
+        // as Alice's balance. Another field is then interpreted as the balance which causes
+        // `BalanceMismatchError`.
```

Comment:
Hmm, this seems like a bad situation. Of course it can't really happen in production because no account will have that much $NEAR, but I still wonder if there's something we can do to avoid this side-effect of the sentinel-based distinction between account types.

Reply:
Will think about it. @jakmeier do you remember your thoughts on why this is the best (or the only viable) solution?

Reply:
We need to be able to distinguish between already existing V1 accounts and new V2 accounts during borsh deserialization. For that, we need a discriminator, which borsh serialized V1 accounts do not have.
If I understand it correctly, the only option to distinguish if the sequence of bytes we are deserializing is V1 or V2 account, is to encounter an invalid value for V1 account.
Possible choices are:
* `amount`
* `locked`
* `code_hash`
* `storage_usage`

`code_hash` cannot be used.
`amount` is better than `locked` because it is the first value.

I found
```
/// StorageUsage is used to count the amount of storage used by a contract.
pub type StorageUsage = u64;
```
If it really represents the amount of storage used by a contract then `amount` looks better, because it is `u128`.
Wdyt?

Reply:
Yeah, I think you're right. This is probably the best solution and we'll leave it as is.